suite-native/firmware/src/hooks/useFirmwareAnalytics.ts

```diff
@@ -0,0 +1,112 @@
+import { useCallback, useEffect, useRef } from 'react';
+
+import { DeviceModelInternal, FirmwareType } from '@trezor/connect';
+import {
+    analytics,
+    EventType,
+    FirmwareUpdatePayload,
+    FirmwareUpdateStartType,
+} from '@suite-native/analytics';
+import { TrezorDevice } from '@suite-common/suite-types';
+import { getFirmwareVersion, getBootloaderVersion } from '@trezor/device-utils';
+
+export const useFirmwareAnalytics = ({
+    device,
+    targetFirmwareType,
+}: {
+    device?: TrezorDevice;
+    targetFirmwareType: FirmwareType;
+}) => {
+    const prepareAnalyticsPayload = useCallback(() => {
+        return {
+            model: device?.features?.internal_model ?? DeviceModelInternal.UNKNOWN,
+            fromBootloaderVersion: getBootloaderVersion(device),
+            fromFwVersion: device?.firmware === 'none' ? 'none' : getFirmwareVersion(device),
+            toFwVersion: targetFirmwareType,
+            fromFwType: (device?.firmwareType || 'none') as FirmwareType | 'none',
+            toFwType: targetFirmwareType,
+        };
+    }, [device, targetFirmwareType]);
+
+    // Use refs to avoid any re-renders because of analytics and to make useCallback dependencies stable
+    // so it won't trigger any useEffect which could interfere with other business logic.
+    const analyticsPayload = useRef<FirmwareUpdatePayload>(prepareAnalyticsPayload());
+    const timeStarted = useRef<number>(Date.now());
+
+    useEffect(() => {
+        analyticsPayload.current = prepareAnalyticsPayload();
+    }, [prepareAnalyticsPayload]);
+
+    const getElapsedTimeInSeconds = useCallback(() => {
+        return Math.floor((Date.now() - timeStarted.current) / 1000);
+    }, []);
+
+    const getAnalyticsPayload = useCallback(() => {
+        return analyticsPayload.current;
+    }, [analyticsPayload]);
+
+    const resetTimeStarted = useCallback(() => {
+        timeStarted.current = Date.now();
+    }, []);
+
+    const handleAnalyticsReportStarted = useCallback(
+        ({ startType }: { startType: FirmwareUpdateStartType }) => {
+            // Q: Maybe we should reset timeStarted only if it's not a retry?
```

Comment:
I believe it's ok to reset the timer, because we report the finished and new start event in that case.

Reply:
[`152ff45` (#16317)](https://github.com/trezor/trezor-suite/pull/16317/commits/152ff457318828fa0ff6697bfe81043ba72476d9)